metals/src/main/scala/scala/meta/internal/metals/SemanticTokensProvider.scala

```diff
@@ -79,15 +83,29 @@ object SemanticTokensProvider {
   def provide(
       nodes: List[Node],
       params: VirtualFileParams,
+      path: AbsolutePath,
       isScala3: Boolean,
+      trees: Trees,
   ): List[Integer] = {
     // no semantic data was available, we can revert to default highlighting
     if (nodes.isEmpty) {
       if (params.uri().toString().isScalaFilename)
         scribe.warn("Could not find semantic tokens for: " + params.uri())
       List.empty[Integer]
     } else {
-      val tokens = getTokens(isScala3, params.text())
+      val tokens = Try(getTokens(isScala3, params.text())) match {
+        case Failure(exception) =>
+          /* Try to get tokens from trees as a fallback,
+           * otherwise this should throw an exception.
+           * If we return empty everything will get unhighlighted.
+           */
+          trees
+            .get(path)
+            .map(_.tokens)
+            .getOrElse(throw exception)
```

Comment:
I still think it better to unhighlight things than throw. I think that if we throw a couple of times VSCode will decide that Metals is erroring and shut down the extension.

Reply:
Sure, this case should be rare enough I guess.